### PR TITLE
ADD CGImage && Update CGColorSpace

### PR DESCRIPTION
**Pull request by LuoZijun**
ADD CGImage Struct.
Update CGColorSpace.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/core-graphics-rs/52)

### CODE REVIEW

**Comment by LuoZijun** [comment]
pls wait this [PR](https://github.com/servo/core-foundation-rs/pull/82) Merge, then retest,thx.

File: src/color_space.rs [hunk 1]
```diff
@@ -8,17 +8,124 @@
...
+//             ColorSpace::DCIP3        => CFString::new("kCGColorSpaceDCIP3"),
+//             ColorSpace::DisplayP3    => CFString::new("kCGColorSpaceDisplayP3")
+//         }
+//     }
+//     pub fn to_CGColorSpaceRef(&self) -> CGColorSpaceRef {
+//         unsafe{
+//             CGColorSpaceCreateWithName(self.to_cfstring.0)
+//         }
+//     }
+// }
```
Comment (pcwalton) [inline src/color_space.rs]: I'd rather just remove the code instead of commenting it out.

File: src/color_space.rs [hunk 2]
```diff
@@ -80,9 +187,52 @@ impl CGColorSpace {
...
+                        obj: CGColorSpaceCreateWithName(CFString::new(s).0)
+                    })
+                },
+                _    => {
+                    Err("error...")
+                }
+            }
+        }
+    }
+}
```
Comment (pcwalton) [inline src/color_space.rs]: Why is this here but `to_string` is commented out?

File: src/display.rs [hunk 3]
```diff
@@ -14,6 +14,7 @@ extern crate core_foundation;
 pub use libc;
 pub use base::{CGError, boolean_t};
 pub use geometry::{CGRect, CGPoint, CGSize};
+use image::{ CGImageRef };
```
Comment (pcwalton) [inline src/display.rs]: Remove `{}`

File: src/image.rs [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+pub enum Pixel {
+    RGB(u8, u8, u8),
+    RGBA(u8, u8, u8, u8),
+    ARGB(u8, u8, u8, u8),
+    BGR(u8, u8, u8),
+    BGRA(u8, u8, u8, u8),
+    ABGR(u8, u8, u8, u8),
+
+    YCbCr(u8, u8, u8),    // YUV420P, 
+    // CMYK()
```
Comment (pcwalton) [inline src/image.rs]: Write `TODO: CMYK` here

File: src/image.rs [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    //         },
+
+    //         ColorSpace::YCbCr => {
+
+    //         },
+    //         _ => {
+    //             println!("Unsupport...");
+    //         }
+    //     }
+    // }
```
Comment (pcwalton) [inline src/image.rs]: Remove this code; don't comment it out

File: src/image.rs [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+
+// Any Documents ?
+// pub const kCGImageAlphaNone: CGImageAlphaInfo               = 0;
+// pub const kCGImageAlphaPremultipliedLast: CGImageAlphaInfo  = 1;
+// pub const kCGImageAlphaPremultipliedFirst: CGImageAlphaInfo = 2;
+// pub const kCGImageAlphaLast: CGImageAlphaInfo               = 3;
+// pub const kCGImageAlphaFirst: CGImageAlphaInfo              = 4;
+// pub const kCGImageAlphaNoneSkipLast: CGImageAlphaInfo       = 5;
+// pub const kCGImageAlphaNoneSkipFirst: CGImageAlphaInfo      = 6;
+// pub const kCGImageAlphaOnly: CGImageAlphaInfo               = 7;
```
Comment (pcwalton) [inline src/image.rs]: Remove this code; don't comment it out

File: src/color_space.rs [hunk 7]
```diff
@@ -8,17 +8,124 @@
...
+use core_foundation::string::{ CFStringRef, CFString };
 use libc::{c_void, size_t};
 use std::mem;
 use std::ptr;
 
+use std::string::ToString;
+use std::str::FromStr;
+
+
+
```
Comment (pcwalton) [inline src/color_space.rs]: Don't use 3 newlines here; use 1

File: src/color_space.rs (hunk 2 above)
Comment (LuoZijun) [inline src/color_space.rs]: @pcwalton I'm sry, ican't understand what's mean about your `to_string `, any detail?

File: src/color_space.rs (hunk 1 above)
Comment (LuoZijun) [inline src/color_space.rs]: @pcwalton agree.

File: src/color_space.rs (hunk 7 above)
Comment (LuoZijun) [inline src/color_space.rs]: agree

File: src/display.rs (hunk 3 above)
Comment (LuoZijun) [inline src/display.rs]: agree

File: src/image.rs (hunk 4 above)
Comment (LuoZijun) [inline src/image.rs]: agree

File: src/image.rs (hunk 5 above)
Comment (LuoZijun) [inline src/image.rs]: agree

File: src/image.rs (hunk 6 above)
Comment (LuoZijun) [inline src/image.rs]: agree

**Comment by application-developer-DA** [comment]
@LuoZijun any updates on this?

File: src/image.rs [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+            let raw_len = CFDataGetLength(cf_data) as usize;
+            // Image size is inconsistent with W*H*D.
+            assert_eq!(self.width()*self.height()*self.pixel_bits(), raw_len*8 );
+            assert_eq!(self.raw_len(), raw_len );
+
+            let data = slice::from_raw_parts(CFDataGetBytePtr(cf_data), raw_len);
+            CFRelease(cf_data as *const c_void);
+            // WARN: Slow.
+            // let v = data.to_vec();
+            data
```
Comment (application-developer-DA) [inline src/image.rs]: Are you sure that the implementation of this function is correct? It seems to me that you're returning a slice pointing to the invalid data. Remember that `slice` does not own the data you passed to it

Basically you created the slice which points to the data you copied from the data provider, and right after that (but before returning the slice itself) you release the data.

**Comment by application-developer-DA** [comment]
@pcwalton is this pull request still needed? `CGImage` has been already merged in master by other pull requests, which added basically the same functionality.

**Comment by jdm** [comment]
I think asking @LuoZijun whether this is still needed makes more sense.

**Comment by application-developer-DA** [comment]
I added some questions to the code review, but received no feedback from him, there were no updates on this pull request since 2016.

CGImage support has been added by https://github.com/servo/core-graphics-rs/pull/79 and then I extended it (initially I wanted to write it, but in parallel another guy made the same pull request, so I just updated mine) with https://github.com/servo/core-graphics-rs/pull/80.

IMO this pull request is obsolete.

**Comment by pcwalton** [comment]
Agreed, this seems obsolete. I'll close this in a few days if there are no objections.
